cmd/prometheus/main.go

```diff
@@ -283,7 +285,21 @@ func main() {
 		remoteStorage.ApplyConfig,
 		webHandler.ApplyConfig,
 		notifier.ApplyConfig,
-		discoveryManager.ApplyConfig,
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.ScrapeConfigs {
+				c[v.JobName] = v.ServiceDiscoveryConfig
+			}
+			return discoveryManagerScrape.ApplyConfig(c)
+		},
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
+				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config pointer as the identifier.
+				c[fmt.Sprintf("%p", v)] = v.ServiceDiscoveryConfig
```

Comment:
Can this be the index of the element in the array?

Reply:
I looked it over and don't think it would be unique enough. There is a chance that when receiving an update from the Discovery to read the wrong config for the target set.

I know it is ugly, but I think less likely to get into a bad sync bug between the configs and the target sets sent from the Discovery manager.